state_tracking.go

```diff
@@ -28,30 +30,53 @@ type StateTrackingNode struct {
 	as     string
 
 	newTracker func() stateTracker
-	groups     map[models.GroupID]*stateTrackingGroup
+
+	groupsMu sync.RWMutex
+	groups   map[models.GroupID]*stateTrackingGroup
 }
 
 func (stn *StateTrackingNode) group(g models.GroupID) (*stateTrackingGroup, error) {
+	stn.groupsMu.RLock()
 	stg := stn.groups[g]
-	if stg == nil {
-		stg = &stateTrackingGroup{}
+	stn.groupsMu.RUnlock()
 
-		var err error
-		stg.Expression, err = stateful.NewExpression(stn.lambda.Expression)
-		if err != nil {
-			return nil, fmt.Errorf("Failed to compile expression: %v", err)
-		}
+	if stg == nil {
+		// Grab the write lock
+		stn.groupsMu.Lock()
+		defer stn.groupsMu.Unlock()
+
+		// Check again now that we have the write lock
+		stg = stn.groups[g]
+		if stg == nil {
```

Comment:
@phemmer You are correct, its just an automatic pattern I follow when using read write mutexes to check the read after acquiring the write lock. It defensive programming. But you are correct since the only place a write lock is acquired is on line 45 then there is technically no need to perform the extra read and nil comparison. But since it is a small cost to pay and is only paid once per new group, I'd rather leave the defensive code in place. That way if later someone sees there is a RWMutex being used they can just grab the write lock and do what they need without having to read all existing locking code.

Thanks for the questions, they are very welcome.